src/main/java/com/artipie/asto/fs/FileStorage.java

```diff
@@ -125,6 +125,10 @@ public FileStorage(final Path path) {
 
     @Override
     public CompletableFuture<Void> save(final Key key, final Content content) {
+        final Path next = this.dir.resolve(key.string());
+        if (!next.normalize().startsWith(next)) {
+            throw new IllegalStateException("Entry path is out of storage.");
+        }
```

Comment:
I'd suggest to wrap this exception in `ArtipieException` or just throw `ArtipieException`

Reply:
@genryxy Ok, in this cas let throw `ArtipieException`